File: src/main/java/repository/TripDBRepo.java
package repository;

import domain.Trip;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class TripDBRepo implements TripRepo {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private final String tableName = "trips";

    public TripDBRepo(Properties props) {
        logger.info("Initializing TripDBRepo with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void save(Trip elem) {
        logger.traceEntry("Saving {}", elem);
        Connection con = dbUtils.getConnection();
        String command = "insert into " + tableName +
                " (tourist_attraction, transport_company, departure_time, price, seats)" +
                " values (?, ?, ?, ?, ?)";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, elem.getTouristAttraction());
            ps.setString(2, elem.getTransportCompany());
            ps.setTime(3, elem.getDepartureTime());
            ps.setFloat(4, elem.getPrice());
            ps.setInt(5, elem.getSeats());
            int result = ps.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Trip elem) {
        deleteById(elem.getId());
    }

    @Override
    public void deleteById(Integer id) {
        logger.traceEntry("Deleting trip with id = {}", id);
        Connection con = dbUtils.getConnection();
        String command = "delete from " + tableName + " where id = ?";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setInt(1, id);
            int result = ps.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Trip elem, Integer id) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "update " + tableName +" set tourist_attraction = ?, transport_company = ?," +
                " departure_time = ?, price = ?, seats = ? where id = ?";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, elem.getTouristAttraction());
            ps.setString(2, elem.getTransportCompany());
            ps.setTime(3, elem.getDepartureTime());
            ps.setFloat(4, elem.getPrice());
            ps.setInt(5, elem.getSeats());
            ps.setInt(6, id);
            int result = ps.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public Trip getById(Integer id) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName + " where id = ?";
        Trip trip = null;
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setInt(1, id);
            ResultSet res = ps.executeQuery();
            res.next();
            String touristAttraction = res.getString("tourist_attraction");
            String transportCompany = res.getString("transport_company");
            Time departureTime = res.getTime("departure_time");
            float price = res.getFloat("price");
            int seats = res.getInt("seats");
            trip = new Trip(id, touristAttraction, transportCompany, departureTime, price, seats);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return trip;
    }

    @Override
    public void clear() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "delete from " + tableName;
        try (PreparedStatement ps = con.prepareStatement(command)) {
            int result = ps.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public Collection<Trip> getAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName;
        List<Trip> trips = new ArrayList<>();
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ResultSet res = ps.executeQuery();
            while (res.next()) {
                int id = res.getInt("id");
                String touristAttraction = res.getString("tourist_attraction");
                String transportCompany = res.getString("transport_company");
                Time departureTime = res.getTime("departure_time");
                float price = res.getFloat("price");
                int seats = res.getInt("seats");
                trips.add(new Trip(id, touristAttraction, transportCompany, departureTime, price, seats));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return trips;
    }

    @Override
    public Collection<Trip> getTouristAttractionTrips(String touristAttraction, Time startTime, Time endTime) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName +
                " where tourist_attraction = ? and departure_time between ? and ?";
        List<Trip> trips = new ArrayList<>();
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, touristAttraction);
            ps.setTime(2, startTime);
            ps.setTime(3, endTime);
            ResultSet res = ps.executeQuery();
            while (res.next()) {
                int id = res.getInt("id");
                String transportCompany = res.getString("transport_company");
                Time departureTime = res.getTime("departure_time");
                float price = res.getFloat("price");
                int seats = res.getInt("seats");
                trips.add(new Trip(id, touristAttraction, transportCompany, departureTime, price, seats));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return trips;
    }
}

File: src/repository/TripDBRepo.cs
using System;
using System.Collections.Generic;
using System.Data;
using AgentiiDeTurism.src.domain;
using log4net;

namespace AgentiiDeTurism.src.repository
{
    class TripDBRepo : ITripRepo
    {
        private static readonly ILog log = LogManager.GetLogger("TripDBRepo");
        private readonly string tableName = "trips";
        IDictionary<string, string> props;
        public TripDBRepo(IDictionary<string, string> props)
        {
            log.Info("Creating TripDBRepo");
            this.props = props;
        }
        public void clear()
        {
            log.Info("Entering clear");
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from " + tableName;
                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Deleted {0} trips", result);
            }
            log.Info("Exiting clear");
        }

        public void delete(Trip elem)
        {
            deleteById(elem.id);
        }

        public void deleteById(int id)
        {
            log.InfoFormat("Entering deleteById with value {0}", id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from " + tableName + " where id = @id";
                IDbDataParameter paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                comm.Parameters.Add(paramID);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Deleted {0} trips", result);
            }
            log.Info("Exiting deleteById");
        }

        public ICollection<Trip> getAll()
        {
            log.Info("Entering getAll");
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from " + tableName;

                using (var dataR = comm.ExecuteReader())
                {
                    List<Trip> trips = new List<Trip>();
                    while (dataR.Read())
                    {
                        int idTrip = dataR.GetInt32(0);
                        string touristAttraction = dataR.GetString(1);
                        string transportCompany = dataR.GetString(2);
                        TimeSpan departureTime = dataR.GetDateTime(3).TimeOfDay;
                        float price = dataR.GetFloat(4);
                        int seats = dataR.GetInt32(5);
                        Trip trip = new Trip(idTrip, touristAttraction, transportCompany, departureTime, price, seats);
                        trips.Add(trip);
                    }
                    log.InfoFormat("Exiting getAll with {0} results", trips.Count);
                    return trips;
                }
            }
        }

        public Trip getById(int id)
        {
            log.InfoFormat("Entering getById with value {0}", id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from " + tableName + " where id = @id";
                IDbDataParameter paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                comm.Parameters.Add(paramID);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idTrip = dataR.GetInt32(0);
                        string touristAttraction = dataR.GetString(1);
                        string transportCompany = dataR.GetString(2);
                        TimeSpan departureTime = dataR.GetDateTime(3).TimeOfDay;
                        float price = dataR.GetFloat(4);
                        int seats = dataR.GetInt32(5);
                        Trip trip = new Trip(idTrip, touristAttraction, transportCompany, departureTime, price, seats);
                        log.InfoFormat("Exiting getById with value {0}", trip);
                        return trip;
                    }
                }
            }
            log.InfoFormat("Exiting getById with value {0}", null);
            return null;
        }

        public ICollection<Trip> getTouristAttractionTrips(string touristAttraction, TimeSpan startTime, TimeSpan endTime)
        {
            log.InfoFormat("Entering getTouristAttractionTrips");
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from " + tableName +
                    " where tourist_attraction = @touristAttraction and departure_time between @startTime and @endTime";

                IDbDataParameter paramTouristAttraction = comm.CreateParameter();
                paramTouristAttraction.ParameterName = "@touristAttraction";
                paramTouristAttraction.Value = touristAttraction;
                comm.Parameters.Add(paramTouristAttraction);
                
                IDbDataParameter paramStartTime = comm.CreateParameter();
                paramStartTime.ParameterName = "@startTime";
                paramStartTime.Value = startTime;
                comm.Parameters.Add(paramStartTime);
                
                IDbDataParameter paramEndTime = comm.CreateParameter();
                paramEndTime.ParameterName = "@endTime";
                paramEndTime.Value = endTime;
                comm.Parameters.Add(paramEndTime);

                using (var dataR = comm.ExecuteReader())
                {
                    List<Trip> trips = new List<Trip>();
                    while (dataR.Read())
                    {
                        int idTrip = dataR.GetInt32(0);
                        string transportCompany = dataR.GetString(2);
                        TimeSpan departureTime = dataR.GetDateTime(3).TimeOfDay;
                        float price = dataR.GetFloat(4);
                        int seats = dataR.GetInt32(5);
                        Trip trip = new Trip(idTrip, touristAttraction, transportCompany, departureTime, price, seats);
                        trips.Add(trip);
                    }
                    log.InfoFormat("Exiting getTouristAttractionTrips with {0} results", trips.Count);
                    return trips;
                }
            }
        }

        public void save(Trip elem)
        {
            log.InfoFormat("Entering save with value {0}", elem);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into " + tableName +
                    " (tourist_attraction, transport_company, departure_time, price, seats)" +
                    "values (@touristAttraction, @transportCompany, @departureTime, @price, @seats)";

                IDbDataParameter paramTouristAttraction = comm.CreateParameter();
                paramTouristAttraction.ParameterName = "@touristAttraction";
                paramTouristAttraction.Value = elem.touristAttraction;
                comm.Parameters.Add(paramTouristAttraction);

                IDbDataParameter paramTransportCompany = comm.CreateParameter();
                paramTransportCompany.ParameterName = "@transportCompany";
                paramTransportCompany.Value = elem.transportCompany;
                comm.Parameters.Add(paramTransportCompany);

                IDbDataParameter paramDepartureTime = comm.CreateParameter();
                paramDepartureTime.ParameterName = "@departureTime";
                paramDepartureTime.Value = elem.departureTime;
                comm.Parameters.Add(paramDepartureTime);

                IDbDataParameter paramPrice = comm.CreateParameter();
                paramPrice.ParameterName = "@price";
                paramPrice.Value = elem.price;
                comm.Parameters.Add(paramPrice);

                IDbDataParameter paramSeats = comm.CreateParameter();
                paramSeats.ParameterName = "@seats";
                paramSeats.Value = elem.seats;
                comm.Parameters.Add(paramSeats);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Saved {0} trips", result);
            }
            log.Info("Exiting save");
        }

        public void update(Trip elem, int id)
        {
            log.InfoFormat("Entering update with values {0}, id = {1}", elem, id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update " + tableName +" set tourist_attraction = @touristAttraction," +
                    " transport_company = @transportCompany, departure_time = @departureTime, price = @price, seats = @seats where id = @id";

                IDbDataParameter paramTouristAttraction = comm.CreateParameter();
                paramTouristAttraction.ParameterName = "@touristAttraction";
                paramTouristAttraction.Value = elem.touristAttraction;
                comm.Parameters.Add(paramTouristAttraction);

                IDbDataParameter paramTransportCompany = comm.CreateParameter();
                paramTransportCompany.ParameterName = "@transportCompany";
                paramTransportCompany.Value = elem.transportCompany;
                comm.Parameters.Add(paramTransportCompany);

                IDbDataParameter paramDepartureTime = comm.CreateParameter();
                paramDepartureTime.ParameterName = "@departureTime";
                paramDepartureTime.Value = elem.departureTime;
                comm.Parameters.Add(paramDepartureTime);

                IDbDataParameter paramPrice = comm.CreateParameter();
                paramPrice.ParameterName = "@price";
                paramPrice.Value = elem.price;
                comm.Parameters.Add(paramPrice);

                IDbDataParameter paramSeats = comm.CreateParameter();
                paramSeats.ParameterName = "@seats";
                paramSeats.Value = elem.seats;
                comm.Parameters.Add(paramSeats);

                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Updated {0} trips", result);
            }
            log.Info("Exiting update");
        }
    }
}
